# Лабораторная работа 1. Язык Python, основные библиотеки для анализа данных.

ФИО: Николаев Владимир Владимирович

Группа: 317

In [1]:
import numpy as np
import pandas as pd

## Часть 1: Python

#### 1. Напишите код, который сформирует список всех натуральных чисел, не превосходящих 100, которые делятся на 7 и не делятся на 5. Нельзя использовать циклы — выполните это задание с помощью list comprehension.

In [2]:
[x for x in range(101) if not(x % 7) and x % 5]

[7, 14, 21, 28, 42, 49, 56, 63, 77, 84, 91, 98]

#### 2. Напишите функцию `sort_tokens(s)`, которая принимает на вход строку, разделяет ее на токены по запятым, сортирует токены с помощью функции `sorted()` и снова соединяет их в одну строку через запятые. Например, для строки `u"ночь,фонарь,улица"` функция должна выдать `u"ночь,улица,фонарь"`. Придумайте несколько входных строк и протестируйте функцию на них.

Обратите внимание, что склеить строки можно с помощью конструкции `'x'.join(...)`. Например:

In [3]:
'! '.join(['i', 'am', 'very', 'excited', ':-)'])

'i! am! very! excited! :-)'

In [4]:
def sort_tokens(s):
    return ",".join(sorted(s.split(",")))

s1 = u"ночь,фонарь,улица"
print(sort_tokens(s1))

s2 = "aaaa,bbbb,cccc"
print(sort_tokens(s2))

s3 = "aaaa,cccc,bbbb"
print(sort_tokens(s3))

ночь,улица,фонарь
aaaa,bbbb,cccc
aaaa,bbbb,cccc


#### 3. Напишите функцию `remove_duplicates(a)`, которая принимает на вход список и возвращает его же, но без дубликатов и в отсортированном виде. Например, для списка `[u"ночь", u"улица", u"фонарь", u"аптека", u"аптека", u"улица", u"фонарь"]` результат должен быть `[u"аптека", u"ночь", u"улица", u"фонарь", ]`.

Подсказка: обратите внимание на структуру `set`!

In [5]:
def remove_duplicates(a):
    return sorted(set(a))

a = [u"ночь", u"улица", u"фонарь", u"аптека", u"аптека", u"улица", u"фонарь"]
remove_duplicates(a)

['аптека', 'ночь', 'улица', 'фонарь']

#### 4. Напишите функцию `word_counts(texts)`, которая принимает на вход список строк, и печатает пары (слово, число вхождений). Считайте, что слова в текстах могут быть разделены только пробелами. Например, для входа `[u'ночь улица фонарь аптека', u'аптека улица фонарь']` должно быть напечатано
>аптека 2

>улица 2

>фонарь 2

>ночь 1

Подсказки:

Чтобы разбить каждый из текстов на слова, воспользуйтесь функцией `map`, которая применяет заданную функцию к каждому элементу списка. Например, если бы нужно было каждый из текстов привести к нижнему регистру, то команда выглядела бы так:

In [6]:
map(lambda x: x.lower(), ["abCDefg", "HIjkL"])

После первого шага вы получите список, состоящий из списков слов. Чтобы объединить их в один список слов, воспользуйтесь функцией `reduce`. Например, просуммировать числа из списка с ее помощью можно так:



In [7]:
reduce(lambda x, y: x + y, [1, 2, 3, 4, 5])

NameError: name 'reduce' is not defined

Для подсчета количества вхождений слов воспользуйтесь структурой `defaultdict` из модуля `collections`. В отличие от обычного `dict`, она не генерирует ошибку при обращении к несуществущему ключу, а создает новую запись, инициализируемую нулевым значением. Например:

In [8]:
from collections import defaultdict
d = defaultdict(int)
print d
d['abc'] += 1
print d
d['xxx']
print d

SyntaxError: Missing parentheses in call to 'print' (<ipython-input-8-95bf4c99c5b8>, line 3)

In [9]:
from functools import reduce
from collections import Counter

def word_counts(texts):
    #Counter is better than defaultdict for this task imho
    counter = Counter(reduce(lambda x, y: x + y, map(lambda x: x.split(" "), texts)))
    for key in counter:
        print(key, counter[key])

s1 = [u'ночь улица фонарь аптека', u'аптека улица фонарь']
word_counts(s1)

аптека 2
ночь 1
фонарь 2
улица 2


#### 5. (бонусное задание) Скачайте английский перевод "Преступления и наказания". Найдите 20 самых популярных слов и 20 слов, которые встречаются только один раз.

Приведите текст к нижнему регистру. Удалите из него все специальные символы и цифры. Разделите текст на слова, считая разделителями все, кроме букв. Обратите внимание на функцию `split` из модуля `re`.

Не забудьте удалить пустые слова:
```
words = filter(lambda x: len(x) > 0, words)
```

In [10]:
from urllib import urlopen
url = "http://www.gutenberg.org/files/2554/2554.txt"
response = urlopen(url)
raw_text = response.read().decode('utf8')

ImportError: cannot import name 'urlopen'

In [11]:
raw_text[1000:1100]

NameError: name 'raw_text' is not defined

In [12]:
import re
import string
f = open("2554.txt", "r")
raw_text = f.readlines()

raw_text = reduce(lambda x,y: x+y, raw_text, "")
raw_text = raw_text.lower()
allow = string.ascii_lowercase + ' '
raw_text = " " + re.sub('[^%s]' % allow, ' ', raw_text)
words = re.split('[^%s]' % string.ascii_lowercase, raw_text)
words = filter(lambda x: len(x) > 0, words)

print("\nMost common words are:")
counter = Counter(words)
for word, count in counter.most_common(20):
    print(word, count)

print("\nLeast common words are:")
for word, count in counter.most_common()[:-21:-1]:
    print(word, count)


Most common words are:
the 8003
and 7042
to 5497
he 4900
a 4668
i 4405
you 4118
of 3928
it 3463
that 3308
in 3254
was 2827
his 2116
at 2082
her 1829
but 1801
not 1775
s 1765
with 1757
she 1695

Least common words are:
stratagem 1
mechanics 1
consoled 1
suggestions 1
squabbles 1
reception 1
foam 1
irrational 1
hail 1
peremptory 1
amateur 1
owning 1
coiners 1
grows 1
submission 1
recklessness 1
nonentity 1
tipsily 1
shil 1
speck 1


1765 слов "s", наверное, от "something's" 

## Часть 2: NumPy и SciPy

#### 1. Повторите 1000 раз следующий эксперимент: сгенерируйте две матрицы размера $10 \times 10$ из стандартного нормального распределения, перемножьте их (как матрицы) и найдите максимальный элемент. Какое среднее значение по экспериментам у максимальных элементов? 95-процентная квантиль?

Для вычисления квантили воспользуйтесь функцией `np.percentile`

In [13]:
avg_max_vector = np.array([])

for it in range(1000):
    m1 = np.random.normal(0, 1, (10, 10))
    m2 = np.random.normal(0, 1, (10, 10))
    avg_max_vector = np.append(avg_max_vector, np.max(np.dot(m1, m2)))
    
print("max: ", np.average(avg_max_vector))
print("95 percentile:", np.percentile(avg_max_vector, 95))

max:  8.48991462837
95 percentile: 12.3514451891


#### 2. Сгенерируйте 1000 чисел из распределения Пуассона с параметром $\lambda = 5$. Сформируйте массив, в котором в i-м элементе будет записано количество сгенерированных чисел, равных i.

Обратите внимание на функцию `np.bincount`

In [14]:
a = np.random.poisson(5, 1000)
counter = np.bincount(a)
print(counter)

[  6  30  83 134 172 181 166 109  59  36  11   7   3   3]


#### 3. Реализуйте функцию, которая принимает на вход numpy-массив целых чисел `a`, и генерирует массив, в котором число `i` встречается `a[i]` раз. Протестируйте на результате из прошлого задания.

Обратите внимание на функцию `np.repeat`

In [15]:
def repeater(a):
    return np.repeat(range(len(a)), a)

print(repeater(a))

[  0   0   0 ..., 999 999 999]


#### 4. Напишите функцию `scale(X)`, которая принимает на вход матрицу и масштабирует каждый ее столбец (вычитает среднее и делит на стандартное отклонение). Убедитесь, что в функции не будет происходить деления на ноль. Протестируйте на каких-нибудь данных.

In [16]:
def scale(X):
    mean = np.mean(X, axis = 0)
    std = np.std(X, axis = 0)
    std[np.abs(std) < 1e-9] = 1
    X = (X - mean) / std
    return X

x = np.array([[1000, 1], [3000, 1]])
print(scale(x))

[[-1.  0.]
 [ 1.  0.]]


#### 5. Сгенерируйте матрицу с элементами из нормального распределения $\mathcal{N}(10, 1)$. Найдите ее:
- определитель
- след
- наименьший и наибольший элементы
- спектральную норму
- норму Фробениуса
- собственные числа
- обратную матрицу

In [17]:
m = np.random.normal(10, 1, (5, 5))
print("Matrix:\n", m)
print("Det: ", np.linalg.det(m))
print("Trace: ", np.trace(m))
print("Min: ", np.min(m), " Max: ", np.max(m), "")
print("Spectral norm: ", np.linalg.norm(m, ord = 2))
print("Frobenius norm: ", np.linalg.norm(m, ord = "fro"))
print("Eigenvalues:\n", np.linalg.eig(m)[0])
print("Inverse matrix:\n", np.linalg.inv(m))

Matrix:
 [[  8.02036802   8.61957099  10.34965565   9.37221754   9.63124136]
 [  9.16201886  11.77164975   9.07152013  10.18305613   9.83395571]
 [ 10.30832248   9.79158691   9.79390269   8.87635844  10.85860878]
 [ 11.4673618    9.54799083  11.37340141  10.61819837   8.96405666]
 [ 10.87062711  10.57055972   9.19742981   8.99858035   9.38746434]]
Det:  113.449267523
Trace:  49.5915831603
Min:  8.02036801895  Max:  11.7716497451 
Spectral norm:  49.373548448
Frobenius norm:  49.5481639707
Eigenvalues:
 [ 49.31246174+0.j          -2.14651281+0.j          -0.36340442+0.j
   1.39451933+1.00231328j   1.39451933-1.00231328j]
Inverse matrix:
 [[-0.56226651  0.05834023  0.4503294   0.357769   -0.34678194]
 [ 0.86227287 -0.42189949 -1.14696534 -0.76977085  1.61906342]
 [ 1.47460899 -1.00207448 -1.44996128 -0.824636    2.00146561]
 [-1.35206862  1.13676048  1.30390113  1.14906985 -2.40913112]
 [-0.46854313  0.29963342  0.94076103  0.15896408 -0.9666347 ]]


## Часть 3: Pandas

#### Ответьте на вопросы о данных по авиарейсам в США за январь-апрель 2008 года.

Данные: http://stat-computing.org/dataexpo/2009/2008.csv.bz2
(обратите внимание, что распаковывать этот файл не обязательно — функция `pandas.read_csv` умеет читать из архивов автоматически)

Описание: http://stat-computing.org/dataexpo/2009/the-data.html

1. Какая из причин отмены рейса (`CancellationCode`) была самой частой? (расшифровки кодов можно найти в описании данных)
2. Найдите среднее, минимальное и максимальное расстояние, пройденное самолетом.
3. Не выглядит ли подозрительным минимальное пройденное расстояние? В какие дни и на каких рейсах оно было? Какое расстояние было пройдено этими же рейсами в другие дни?
4. Из какого аэропорта было произведено больше всего вылетов? В каком городе он находится?
5. Найдите для каждого аэропорта среднее время полета (`AirTime`) по всем вылетевшим из него рейсам. Какой аэропорт имеет наибольшее значение этого показателя?
6. Найдите аэропорт, у которого наибольшая доля задержанных (`DepDelay > 0`) рейсов. Исключите при этом из рассмотрения аэропорты, из которых было отправлено меньше 1000 рейсов (используйте функцию `filter` после `groupby`).

In [18]:
data = pd.read_csv("2008.csv.bz2")

In [19]:
print(data["CancellationCode"].value_counts(dropna=False))
print(len(data))

NaN    6872294
B        54904
A        54330
C        28188
D           12
dtype: int64
7009728


Самой частой причиной были погодные условия (B - weather).

In [20]:
data["Distance"].describe()

count    7009728.000000
mean         726.387029
std          562.101803
min           11.000000
25%          325.000000
50%          581.000000
75%          954.000000
max         4962.000000
Name: Distance, dtype: float64

* Среднее: 726.38 миль
* Минимальное: 11 миль
* Максимальное: 4962 миль

In [21]:
subset = data[(data["Distance"] - 11) < 0.5]
subset[["Year","Month","DayofMonth","DayOfWeek", "FlightNum", "UniqueCarrier"]]

Year  Month  DayofMonth  DayOfWeek  FlightNum UniqueCarrier
2547298  2008      5          15          4       4988            OH
4392215  2008      8          10          7       5572            OH

In [22]:
print("Flight 4988 by carrier 'OH':")
print(data[np.logical_and(data["FlightNum"] == 4988, data["UniqueCarrier"] == "OH")]["Distance"].describe())
print("\nFlight 5572 by carrier 'OH':")
print(data[np.logical_and(data["FlightNum"] == 5572, data["UniqueCarrier"] == "OH")]["Distance"].describe())

Flight 4988 by carrier 'OH':
count    132.000000
mean     822.174242
std       95.849476
min       11.000000
25%      834.000000
50%      834.000000
75%      834.000000
max      834.000000
Name: Distance, dtype: float64

Flight 5572 by carrier 'OH':
count    244.000000
mean     541.815574
std       34.121829
min       11.000000
25%      544.000000
50%      544.000000
75%      544.000000
max      544.000000
Name: Distance, dtype: float64


Минимальное расстояние в 11 миль было пройдено рейсами 4988 (15/5/2008) и  5572 (10/8/2008).
Среднее растояние на этих рейсах равно 822.17 и 541.82 милям соответственно.

In [23]:
data["Origin"].value_counts()[:1]

ATL    414513
dtype: int64

Больше всего вылетов было произведено из аэропорта ATL ("William B Hartsfield-Atlanta Intl", Atlanta)

In [24]:
mean_air_time = data.groupby("Origin")["AirTime"].mean()
print(np.argmax(mean_air_time))

SJU


Наибольшее среднее время полета имеет аэропорт SJU ("Luis Munoz Marin International",San Juan)

In [25]:
data = data.groupby("Origin").filter(lambda x: len(x) >= 1000)

In [26]:
data_delayed = data[data["DepDelay"] > 0].groupby("Origin").size()

In [27]:
data_all = data.groupby("Origin").size()

In [28]:
df = pd.concat([data_all, data_delayed], axis = 1)

In [29]:
np.argmax(df[1] / df[0])

'DAL'

DAL ("Dallas Love", Dallas)- аэропорт с наибольшей долей задержанных рейсов.